### DISCUSSION TITLE
Is QBluetoothPermission() of PySide 6.6.* supported by latest Pyinstaller 6.4.0 on MacOs 13.0?

### DISCUSSION THREAD

**Discussion by Rida-Shamasneh** [pyinstaller/pyinstaller#8347] [Help] [answered]
I am trying to get rid of the following message which I get whenever I try to discover BLE devices:
`Use of Bluetooth LE requires explicitly requested permissions.`

I tried today to request bluetooth permissions expliclity but I am always getting "Denied!" !!

```
def __request_bluetooth_permissions_explicitly(self):
        ble = QBluetoothPermission()
        ble.setCommunicationModes(QBluetoothPermission.Access)
        status = QApplication.instance().checkPermission(ble)
        if status == Qt.PermissionStatus.Undetermined:
            print("Undetermined!")
            QApplication.instance().requestPermission(ble, self, self.__request)
            return
        elif status == Qt.PermissionStatus.Denied:
            print("Denied!")
        elif status == Qt.PermissionStatus.Granted:
            print("Granted!")

self.__discovery_agent.DiscoveryMethod = QBluetoothDeviceDiscoveryAgent.LowEnergyMethod
self.__discovery_agent.start()
```

- PySide6.6.2
- Pyinstaller 6.4.0 (the latest)
- info_plist file including "NSBluetoothAlwaysUsageDescription" string.
- MacOs 13.0

I wonder if Pyinstaller supports this?

**Comment by bwoodsend** [comment]
<details><summary>So that we're all on the same page, a quick and dirty reproducer:</summary>

```python
from PyQt6 import QtWidgets, QtCore

app = QtWidgets.QApplication([])
ble = QtCore.QBluetoothPermission()
ble.setCommunicationModes(QtCore.QBluetoothPermission.CommunicationMode.Access)

app.requestPermission(ble, print)

status = app.checkPermission(ble)
if status == QtCore.Qt.PermissionStatus.Undetermined:
    message = "Undetermined!"
elif status == QtCore.Qt.PermissionStatus.Denied:
    message = "Denied!"
elif status == QtCore.Qt.PermissionStatus.Granted:
    message = "Granted!"

window = QtWidgets.QLabel(message)
window.show()
app.exec()
```

</details>

If I put the following in the spec file:
```python
app = BUNDLE(
    ...
    info_plist={
      "NSBluetoothAlwaysUsageDescription": "I want your bluetooth and I want it now",
   },
)
```
then I get a pop up that looks like:
<img width="372" alt="image" src="https://github.com/pyinstaller/pyinstaller/assets/30940778/c2b6656b-2603-4a90-bed2-71103226a548">
and the *Undetermined* response but if I hit OK, then run again I get *Granted*.

**Reply by bwoodsend** [reply]
Actually, here's a more satisfying example which removes the need to restart the application:
```python
from PyQt6 import QtWidgets, QtCore

app = QtWidgets.QApplication([])
ble = QtCore.QBluetoothPermission()
ble.setCommunicationModes(QtCore.QBluetoothPermission.CommunicationMode.Access)


def check_status():
    status = app.checkPermission(ble)
    if status == QtCore.Qt.PermissionStatus.Undetermined:
        message = "Undetermined!"
    elif status == QtCore.Qt.PermissionStatus.Denied:
        message = "Denied!"
    elif status == QtCore.Qt.PermissionStatus.Granted:
        message = "Granted!"
    return message


window = QtWidgets.QLabel(check_status())
app.requestPermission(ble, lambda _: window.setText(check_status()))

window.show()
app.exec()
```

**Reply by bwoodsend** [reply]
You might want to check your _system settings_ -> _privacy and security_ -> _bluetooth_ and make sure that your application (or terminal if you're running from terminal) isn't listed and disallowed.

**Reply by Rida-Shamasneh** [reply]
Thanks @bwoodsend for the quick response.
Would it work with PySide6.6.2?

**Reply by Rida-Shamasneh** [reply]
I tested your logic with PyQt6.6.1 and worked as a charm.
It would be great if you can help with a PySide6 example.

**Reply by Rida-Shamasneh** [reply]
I tested the following PySide6 example that does the same as yours; But -unfortunately- it did not work:

```
from PySide6 import QtWidgets, QtCore
from PySide6.QtWidgets import QApplication

app = QtWidgets.QApplication([])
ble = QtCore.QBluetoothPermission()
ble.setCommunicationModes(QtCore.QBluetoothPermission.CommunicationMode.Access)


class Checker:
    @staticmethod
    def __func__():
        status = QApplication.instance().checkPermission(ble)
        message = ""
        if status == QtCore.Qt.PermissionStatus.Undetermined:
            message = "Undetermined!"
        elif status == QtCore.Qt.PermissionStatus.Denied:
            message = "Denied!"
        elif status == QtCore.Qt.PermissionStatus.Granted:
            message = "Granted!"
        return message


window = QtWidgets.QLabel(Checker.__func__())
window.setFixedSize(300, 300)
app.requestPermission(ble, window, Checker)
window.setText(Checker.__func__())
print(Checker.__func__())

window.show()
app.exec()
```

PS - I am not fully sure that my example code is 100% correct; because there is a slight change in the interface between pyqt6.6.1 and pyside6.6.*

**Reply by Rida-Shamasneh** [reply, marked as answer]
@bwoodsend 

It seems issue is in PySide6.6.* according to the following ticket I profiled to pyside devs:
https://bugreports.qt.io/browse/PYSIDE-2474

"_We are aware of the issue and a fix is already in place. It should be available from 6.7. See https://bugreports.qt.io/browse/PYSIDE-2468, but you will have to use pyside6-deploy or Nuitka._"

**Reply by rokm** [reply]
Looks like PySide6 is currently not including/linking plugins for requesting permissions on macOS, and that will be remedied by https://codereview.qt-project.org/c/pyside/pyside-setup/+/530714 (referenced in https://bugreports.qt.io/browse/PYSIDE-2468) in `PySide6` v6.7 (and presumably v6.6.3, if that happens to be released before v6.7). So when new version of `PySide6` is released, it should work if you add `NSBluetoothAlwaysUsageDescription` to `Info.plist` as in examples above.

**Reply by Rida-Shamasneh** [reply]
@rokm 
I downloaded and installed [pyside6.7-dev version captured today 8-March-2024](https://download.qt.io/snapshots/ci/pyside/dev/0b43807deee7ebfe0d5a2fd79e07a6cbab569970/) and I was able to get a "Granted!" result.
However; when I built my tool with this dev version (with Pyinstaller 6.4.0); I noticed two issues:
  - Some issues during BLE devices discovery procedure. I was able to tweak it a little bit, and I was able to discover my BLE device.
  - But when I tried to connect; the tool crashed ( a SEG Fault) :( 
  Anyway; These are different issues that probably have nothing to do with the reported issue here; and might have been occurred because I tried a dirty-dev build. Anyway; I do not think it would be part of v6.6.3 according to PySide dev comment shared in my comment above.

**Reply by Rida-Shamasneh** [reply]
@rokm 

PySide6.7.0 was released a couple of days ago. I tried to bundle my pyside6 via latest pyinstaller v6.5.0 .. but unfortunately seems not yet supported by QApplication.instance().requestPermission, am I right?

**Reply by Rida-Shamasneh** [reply]
I just noticed something strange:
The bundled app requests Bluetooth permissions successfully if the app does not exist in the list of apps under `Privacy & Security  / Bluetooth` ,,, but if it is already there and disabled, then I can not request permissions .... 

In other words; my question is:
* Shall this API be able to request BT permissions in all cases, especially if app is listed under `Privacy & Security  / Bluetooth` but turned off there?

**Reply by rokm** [reply]
> * Shall this API be able to request BT permissions in all cases, especially if app is listed under `Privacy & Security  / Bluetooth` but turned off there?

It probably shouldn't, otherwise your app would be able to nag the user for those permissions each time they run the app, until they finally give in and grant it the permissions.
